fs/nfs/nfs4proc.c

```diff
@@ -7771,6 +7775,475 @@ static bool nfs4_xattr_list_nfs4_sacl(struct dentry *dentry)
 	return nfs4_server_supports_acls(NFS_SB(dentry->d_sb), NFS4ACL_SACL);
 }
 
+#if CONFIG_TRUENAS
+/*
+ * We will publish the DACL thru NA41_NAME ("system.nfs4_acl_xdr")
+ *
+ * First some support functions.  These may be similar to upstream
+ * functions in nfsd.
+ */
+
+/*
+ * return the size of the struct nfs4_acl required to represent an acl
+ * with @entries entries.
+ */
+static int nfs4_acl_bytes(int entries)
+{
+	return sizeof(struct nfs4_acl) + entries * sizeof(struct nfs4_ace);
+}
+
+static struct {
+	char *string;
+	int   stringlen;
+	int type;
+} s2t_map[] = {
+	{
+		.string    = "OWNER@",
+		.stringlen = sizeof("OWNER@") - 1,
+		.type      = NFS4_ACL_WHO_OWNER,
+	},
+	{
+		.string    = "GROUP@",
+		.stringlen = sizeof("GROUP@") - 1,
+		.type      = NFS4_ACL_WHO_GROUP,
+	},
+	{
+		.string    = "EVERYONE@",
+		.stringlen = sizeof("EVERYONE@") - 1,
+		.type      = NFS4_ACL_WHO_EVERYONE,
+	},
+};
+
+static int nfs4_acl_get_whotype(char *p, u32 len)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(s2t_map); i++) {
+		if (s2t_map[i].stringlen == len &&
+		    (memcmp(s2t_map[i].string, p, len) == 0))
+			return s2t_map[i].type;
+	}
+	return NFS4_ACL_WHO_NAMED;
+}
+
+/*
+ * XDR encode the ace who into the supplied buffer.
+ *
+ * Return the number of bytes (including the leading nbytes and
+ * any padding necessary)
+ */
+static int nfs4_ace_encode_who(struct nfs4_ace *ace, char *buf)
+{
+	u32 *p = (u32 *)buf;
+	int count = -EINVAL;
+	const char *pstr = NULL;
+	char idbuf[11];
+	int total_bytes, pad_count;
+
+	/*
+	 * First find the size of the string (into count)
+	 *
+	 * Will also have some side-effects which will be used below:
+	 * either idbuf will be populated (for NFS4_ACL_WHO_NAMED) or
+	 * pstr will point into s2t_map.
+	 */
+	if (ace->whotype == NFS4_ACL_WHO_NAMED) {
+		u32 id;
+
+		if (ace->flag & NFS4_ACE_IDENTIFIER_GROUP)
+			id = from_kgid_munged(&init_user_ns, ace->who_gid);
+		else
+			id = from_kuid_munged(&init_user_ns, ace->who_uid);
```

Comment:
This particular module is nfs client, but the issue remains.  However, since the PR is to allow **us** to use our ZFS acl manipulation tools (i.e. not in a container) it should be ok.

Reply:
Can somebody try (if it is allowed) to use NFS client inside a container and get wrong idmap?